### ISSUE TITLE
Forgetting to detach() a DTensor graph output that does not need computed gradients leads to bad error message

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#172556]
If you look at this example:
```
import torch
import torch.distributed as dist
from torch.distributed._tensor import DTensor, DeviceMesh, Shard, distribute_tensor
from torch.testing._internal.distributed.fake_pg import FakeStore
# Initialize with fake process group
store = FakeStore()
dist.init_process_group(backend="fake", rank=0, world_size=2, store=store)
# Create a device mesh
device_mesh = DeviceMesh("cuda", [0, 1])
# Create two tensors and convert to DTensors
t1 = torch.randn(4, 4, device="cuda")
dt1 = distribute_tensor(t1, device_mesh, [Shard(0)])
dt1.detach_().requires_grad_()
@torch.compile(backend="aot_eager")
def f(x):
    # This fails
    return x.sin(), x.cos()
    # Replacing with this line works
    # return x.sin(), x.cos().detach()
out1, out2 = f(dt1)
out1.sum().backward()
dist.destroy_process_group()
```

You can see that the example does something inefficient: `f` returns two DTensor outputs, and we only backprop against one of them (`out1`). Both outputs require_grad, though, which means that even after running `out1.sum().backward()`, any activations stored in the graph that were part of `out2`'s compute will stay live in memory (even though, in this example, we never actually cared about backpropping `out2`).

In eager, this just leads to wasted memory. Under compile, though, this becomes a hard error:
```
[rank0]:     AOTDispatchAutograd.process_runtime_tangent(
[rank0]:   File "/data/users/hirsheybar/new2/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 2179, in process_runtime_tangent
[rank0]:     raise RuntimeError(
[rank0]: RuntimeError:
[rank0]: During the backward, we encountered a tensor subclass where we guessed its
[rank0]: metadata incorrectly.

[rank0]: Expected metadata: (DTensorSpec(mesh=DeviceMesh((2,), 'cuda', stride=(1,)), placements=(Shard(dim=0),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32), shard_order=(ShardOrderEntry(tensor_dim=0, mesh_dims=(0,)),)), False), expected type: <class 'torch.distributed.tensor.DTensor'>

[rank0]: Runtime metadata: None, runtime type: <class 'torch.Tensor'>

[rank0]: shape: torch.Size([4, 4])
[rank0]: To fix this, your tensor subclass must implement the dunder method __force_to_same_metadata__.
```

The issue is as follows:

(1) torch.compile traces out a backward graph ahead-of-time, before the user actually calls `.backward()`
(2) in order to do this, we need to create some backward inputs to trace the backward portion of the graph with
(3) we make the assumption that for each forward output that is a DTensor, its corresponding backward input will also be a DTensor
(4) if you call `.backward()` on only a subset of your outputs that require grad, though, then the autograd will replace any "unused" gradients with plain tensors of zeros. So at runtime, the value for `grad_out2` that gets plumbed into the backward graph is a plain tensor of zeros. We had no way of correctly guessing that this might happen at trace time.

This is part of the larger issue of "by tracing the backward ahead of time, we can guess tangent metadata incorrectly". And there are two pretty reasonable workarounds for this case:

(1) just ensure that you detach() any graph outputs that you don't plan on computing gradients for. This will be better for memory usage anyway

(2) if you are willing to compile a single fwd-loss-bwd graph, you will also not run into this problem

The main reason for this github issue is that the error message right now is not very good. Maybe we should consider making it more actionable?

cc @chauhang @penguinwu @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx @bobrenjc93

**Comment by arkadip-maitra** [comment]
Added error message pr and made it conditional on the type of error
